rfcs/experimental/flow_graph_wait_try_put_and_wait/README.md

```diff
@@ -0,0 +1,604 @@
+
+``message_metainfo`` is broadcasted through the graph together with the message itself. The actual value and the type of the message can be changed during the computation, but the stored 
+metainformation is preserved.
+
+The reference counter in the ``wait_context`` inside of the ``message_metainfo`` is increased in the following cases:
+* The task associated with the computations of the ``try_put_and_wait`` input message is created (*).
+* The item associated with the computations of the desired input is stored in the buffering node or in the internal buffer of the other node type.
+* When the ``continue_msg`` with the non-empty associated waiter is received by ``continue_node``. In this case, the node buffers the ``message_metainfo``s received from each
+  predecessor and increases the underlying reference counter to prolonge the wait until signals from each predecessor would be received. 
+
+The reference counter is decreased when:
+* The task associated with the computations of the ``try_put_and_wait`` input is finalized. If the output of the task should be propagated to the successors of the node, it is done
+  from the task body and hence the reference counter would be decreased only after creating the task for each successor and increasing the reference counter for them.
+* The item associated with the computations is taken from the buffering node or from the internal buffer.
+* When the desired number of signals from the predecessors was received by the ``continue_node``. This case is equivalent to retrieving the set of buffered ``message_metainfo``s received previously
+  from each predecessor. 
+
+``message_metainfo`` class may be reused in the future to support additional use-cases when it is required to push additional information about the input message through the graph. E.g. supporting
+priorities for single messages - the corresponding priority tag can be assigned as part of the metainfo.
+
+Metainformation class supports containing multiple reference counters at the same time to support joining messages with different associated reference counters in the ``join_node``. 
+See [separate section](#details-about-metainformation-class) for more details.
+
+Implementation-wise, processing the metainformation is exposed by adding new internal virtual member functions to various Flow Graph instance:
+
+``` cpp
+template <typename T>
+class receiver {
+protected:
+    virtual bool try_put_task(const T& t) = 0; // Existing API
+    virtual bool try_put_task(const T& t, const message_metainfo& info) = 0; // New API
+};
+```
+
+For each particular implementation of ``receiver``, the ``try_put_task`` performs ab action that is required by the corresponding Flow Graph node. 
+
+It may buffer both ``t`` and ``metainfo`` or broadcast the result and the ``info`` to the successors of the node. 
+
+The existing API ``try_put_task(const T& t)`` can reuse the new one with the empty metainfo object.
```

Comment:
Would not that however increase overheads for programs that do not use `try_put_and_wait()`?

Reply:
I have made an initial analysis before the feature was released and did not see any performance degradations (at least on our examples and some simple benchmarks). Currently, the empty metainfo is represented as an empty `std::forward_list` that is almost identical to the single nullptr.

I have added a performance analysis section to the exit criteria that should cover both performance improvement while `try_put_and_wait` is used and absence of degradations when it is not used.